### PR TITLE
Add replace and replace_if algorithms

### PR DESCRIPTION
**Pull request by mkurdej**
replace: replaces all occurrences of a given type in the sequence by a new type.
replace_if: replaces all types fullfiling a given predicate in the sequence by a new type.

### CODE REVIEW

**Comment by mkurdej** [comment]
llvm.org is down, that's why the build fails.

File: brigand/algorithms/detail/replace.hpp [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  (See accompanying file LICENSE.md or copy at http://boost.org/LICENSE_1_0.txt)
+=================================================================================================**/
+#pragma once
+
+#include <brigand/functions/lambda/apply.hpp>
+#include <brigand/types/bool.hpp>
+#include <type_traits>
+
+#ifdef _MSC_VER
+#define BRIGAND_DEDUCED_TEMPLATE template
```
Comment (jfalcou) [inline brigand/algorithms/detail/replace.hpp]: I smell MSVC2013 bullshit here, am I right ?

File: brigand/algorithms/detail/replace.hpp [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+struct replace_if_impl
+{
+    using type = Sequence;
+};
+
+template <typename Predicate, typename T, typename NewType>
+struct predicate_test
+{
+    using type = typename std::conditional<
+        /*B=*/brigand::BRIGAND_DEDUCED_TEMPLATE apply<Predicate, T>::type::value,
```
Comment (jfalcou) [inline brigand/algorithms/detail/replace.hpp]: apply is a regular metafunction, why the need of template here ?

File: brigand/algorithms/detail/replace.hpp (hunk 1 above)
Comment (mkurdej) [inline brigand/algorithms/detail/replace.hpp]: Yeah, MSVC2013, with 2015 it's fine... Completely illogical for me.

Error from MSVC2013:
```
3>C:\workspace\brigand\brigand/algorithms/detail/replace.hpp(35): error C2825: 'brigand::apply': must be a class or namespace when followed by '::'
3>          C:\workspace\brigand\brigand/algorithms/detail/replace.hpp(38) : see reference to class template instantiation 'brigand::detail::predicate_test<Predicate,T,NewType>' being compiled
```

File: brigand/algorithms/detail/replace.hpp (hunk 2 above)
Comment (mkurdej) [inline brigand/algorithms/detail/replace.hpp]: No idea.

**Pull request merged**